File: N10003/main.py
import sys


dp = [[0] * 52 for i in range(52)]


def solve(cuts_input, l):
    cuts = [0] + cuts_input + [l]
    n = len(cuts) - 1

    for i in range(n):
        dp[i][i] = 0

    for col in range(1, n):
        for row in range(col - 1, -1, -1):
            dp[row][col] = (min(dp[row][x] + dp[x + 1][col]
                                for x in range(row, col))
                            + cuts[col + 1]
                            - cuts[row])

    return dp[0][n - 1]


def main(file):
    res = []
    while True:
        l = int(file.readline())
        if not l:
            break
        n = int(file.readline())
        sol = solve([int(x) for x in file.readline().split()], l)
        res.append('The minimum cutting is {}.\n'.format(sol))
    return res


if __name__ == '__main__':
    print(''.join(main(sys.stdin)), end='')

File: N10003/main.c
#include <stdio.h>
#include <limits.h>

int dp[52][52];

int solve(int cuts[], int size, int l)
{
    int row, col, n;
    int x, min, value;

    n = size + 1;
    cuts[0] = 0;
    cuts[n] = l;

    for (x = 0; x < n; x++) {
        dp[x][x] = 0;
    }

    for (col = 1; col < n; col++) {
        for (row = col - 1; row >= 0; row--) {
            min = INT_MAX;
            for (x = row; x < col; x++) {
                value = dp[row][x] + dp[x + 1][col];
                if (value < min) {
                    min = value;
                }
            }
            dp[row][col] = min + cuts[col + 1] - cuts[row];
        }
    }

    return dp[0][n - 1];
}

int main()
{
    int l, n, cut;
    int cuts[52];

    while (1) {
        scanf("%d\n", &l);
        if (l == 0) {
            break;
        }
        scanf("%d\n", &n);
        for (cut = 1; cut <= n; cut++) {
            scanf("%d", &cuts[cut]);
        }
        printf("The minimum cutting is %d.\n", solve(cuts, n, l));
    }
}